[CI] Publish binaries for Aspire.Cli

radical opened the pull request:
Generate self contained single file binaries for Aspire.Cli for the following platforms:

- linux-arm64
- linux-x64
- linux-musl-x64
- osx-arm64
- osx-x64
- win-arm64
- win-x64
- win-x86

Produces archives named like `aspire-cli-linux-x64-9.4.0-preview.1.25316.11.tar.gz` in `artifacts/packages/$(CONFIG)/Shipping`.

Contributes to https://github.com/dotnet/aspire/issues/9638

davidfowl commented:
How big is it

joperezr commented on eng/clipack/Common.projitems:
```diff
@@ -0,0 +1,40 @@
+<Project>
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
```
I would have expected using the NoBuild targets sdk instead. Is there a reason why we need this one? Without it we can probably avoid having to set a target Framework.

joperezr commented on eng/clipack/Common.projitems:
```diff
@@ -0,0 +1,40 @@
+      <AdditionalProperties Include="RuntimeIdentifier=$(CliRuntime)" />
+      <AdditionalProperties Include="PublishSingleFile=true" />
+      <AdditionalProperties Include="SelfContained=true" />
+      <!-- <AdditionalProperties Include="GenerateDocumenationFile=false" /> -->
```
Is this needed?

joperezr commented on eng/clipack/Common.projitems:
```diff
@@ -0,0 +1,40 @@
+      Text="No files were found to pack. Ensure that the project $(MSBuildProjectFullPath) has a publish target defined." />
+
+    <MakeDir Directories="$(CliPublishedArtifactsOutputDir)/$(CliRuntime)" />
+    <ZipDirectory
```
I know this is how we do it for the dashboard, but since this will be customer facing we should change it such that only Windows builds use .zip files, but Linux artifacts should generate .tar files instead. Most Linux distros come with utilities to extract tars, but not zips so we should avoid having this as a requirement.

radical replied:
Hm I can't reference System.IO.Compression and System.Formats.Tar from the inline task to make this work. I'll create a new tasks project for this, and use that.

radical committed:
[CI] Generate and publish Cli as single file executable

- This generates zip files for non-linux platforms. And `.tar.gz` for
linux.
- And uses Arcade blob publishing to publish the generated archives

joperezr commented on eng/Publishing.props:
```diff
@@ -54,6 +56,11 @@
         <PublishFlatContainer>true</PublishFlatContainer>
         <RelativeBlobPath>$(_UploadPathRoot)/$(_PackageVersion)/%(Filename)%(Extension)</RelativeBlobPath>
       </ItemsToPushToBlobFeed>
+      <ItemsToPushToBlobFeed Include="@(_CliFilesToPublish)">
```
Totally ok if you prefer to do that as a follow-up PR

radical replied:
Yeah, that would have to be in a follow-up.

joperezr commented on eng/clipack/Aspire.Cli.linux-arm64.csproj:
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <PropertyGroup>
+    <CliRuntime>linux-arm64</CliRuntime>
```
No need to block your PR for this, but we may need to add a linux-musl build as well. Thoughts @mitchdenny @DamianEdwards @davidfowl

mitchdenny replied:
I think that @danegsta was adding in/has added in the musl support on the DCP side too.

joperezr replied:
This is a new project though, so he wouldn't have done it for this one.

mitchdenny replied:
Well there is no point shipping musl if the whole stack doesn't support it :)

joperezr commented on eng/clipack/Common.projitems:
```diff
@@ -0,0 +1,59 @@
+
+    <Exec
+      Condition="$(CliRuntime.StartsWith('linux-'))"
+      Command="pwsh &quot;$(_CreateTarGzScriptPath)&quot; @(_CreateTarGzArguments, ' ')"
```
This now means there is a new prereq in the build as I don't believe pwsh was required before, correct?

Was there nothing in the SDK that does this naitvely? Might also be worth checking with dotnet monitor CLI tool as they produce tars as part of their build so we should also see what they do. TL:DR is that if we can simplify and get to the same results without having to introduce new pre-reqs and custom scripts, it'd be better.

radical replied:
re:pwsh, we use that in multiple places for running tests on github actions.

re:sdk, there is a https://github.com/dotnet/sdk/blob/ad24473f7f2ae16eaee4cbd28eb6897d00df32f3/src/Tasks/sdk-tasks/TarGzFileCreateFromDirectory.cs#L8 which isn't being surfaced in a nuget. But Eric suggested https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.Tar which I will be trying. Still in draft :D

radical replied:
Excellent find! There seems to be `[Microsoft.DotNet.Build.Tasks.Archives](https://github.com/dotnet/arcade/tree/d9d02d858b71562509f72ea84409853e4deffc8f/src/Microsoft.DotNet.Build.Tasks.Archives)` nuget which IIUC supports this. I will try it out.

radical committed:
Use Microsoft.DotNet.Build.Tasks.Archives to create archives

radical commented:
WIP, not ready for review yet.

radical commented:
next:
- [ ] Confirm that the blobs got published, and are accessible
- [ ] How can we use notargets sdk with `Microsoft.DotNet.Build.Tasks.Archives`

radical committed:
clipack: Use Microsoft.Build.NoTargets

radical committed:
add Aspire.Cli.linux-musl-x64.csproj

radical committed:
Merge remote-tracking branch 'origin/main' into cli-single

radical committed:
cleanup

radical committed:
Fix aspire.cli linux-musl build

joperezr reviewed (approved):
:shipit:

joperezr commented on eng/clipack/Common.projitems:
```diff
@@ -0,0 +1,33 @@
+    <PackageReference Include="Microsoft.DotNet.Build.Tasks.Archives" />
+  </ItemGroup>
+
+  <Target Name="PublishToDisk">
```
NIT: don't change this now, but might be good to include in your next PR a small comment here saying that the Microsoft.DotNet.Build.Tasks.Archives targets will automatically invoke this target when the project is published so that others in the future understand how this is plugged in.
